### PR TITLE
Fix 640, 642 - parameter intellisense for static parameters to methods

### PR DESCRIPTION
**Pull request by dsyme**
This is an initial stab at fixing https://github.com/Microsoft/visualfsharp/issues/640 and https://github.com/Microsoft/visualfsharp/issues/642.  I believe it's the correct fix but it needs testing added.  I'm pushing it to give visibility particularly for @dmitry-a-Morozov

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
How to test it? I downloaded the release.zip from AppVeyor and found net40 folder in it, which looks like this:
image
What I should do next? Will my normal VS installation be messed in such a way that it must be reinstalled after the testing?

**Comment by 7sharp9** [comment]
Maybe a good idea to add this to FCS, that would allow a quick feedback test?

**Comment by KevinRansom** [comment]
@dsyme Where are we at with this PR is it still WIP?

**Comment by dsyme** [comment]
Yes still WIP

**Comment by dsyme** [comment]
This PR is now a bunch of fixes to tooling support for static method parameters, it's code complete and I've done manual testing, but we need to add automated tests.

Basically autocomplete, quick info and parameter info support for static method parameters just wasn't done at all in F# 4.0.  That's my fault, I just focused on the core language feature and forgot to test the tooling support beyond the very simplest of cases, and the QA done didn't catch it.  I think if we'd had a proper RFC process in place we would have caught it as we would have noted the need for tooling support in the RFC, and hence tested it.

That said, it was actually hard to add tooling support in early 2015 when we did this feature: the repo just was a bit raw w.r.t. testing in this area.  Now that the FSharp.Compiler.Service tests are in the repo and all the VS tests are passing under CI, it's a lot easier to implement, trial and validate tooling support, and to have confidence that we're not busting anything in the process.

In any case, the test cases I've covered manually were all testing the "CreateCommand" magic method of https://github.com/fsprojects/FSharp.Data.SqlClient/, 

```fsharp
DB.CreateCommand  // TEST: intellisense when typing "<"
DB.CreateCommand<CMD>( ) // TEST: param info on "("
DB.CreateCommand<" SELECT 17" >( ) // TEST: param info on "("
DB.CreateCommand<" SELECT 17" >( (*$*) ) // TEST: meth info on ctrl-alt-space at $
DB.CreateCommand<" SELECT 17" >( null (*$*) ) // TEST: param info on "," at $
DB.CreateCommand   // TEST: should give proper error message if unapplied
DB.CreateCommand< > // TEST: intellisense when typing "<"
DB.CreateCommand< (*$*) > // TEST: param info when typing ctrl-alt-space at $
DB.CreateCommand<" SELECT 17" (*$*) > // TEST: param info on Ctrl-alt-space at $
DB.CreateCommand<" SELECT 17" (*$*) >(  ) // TEST: param info on Ctrl-alt-space at $
DB.CreateCommand<" SELECT 17", (*$ *) >(  ) // TEST: param info on Ctrl-alt-space at $
DB.CreateCommand<" SELECT 17" >(  (*$*) ) // TEST: no assert on Ctrl-space at $
```

**Comment by dsyme** [comment]
Additionally, this PR fixes a case in the F# language design where you couldn't use a generic static method which has been applied to type parameters as a first class function value, e.g.

```fsharo

type C() = 
    static member M<'T>() = Unchecked.defaultof<'T>

let f = C.M<int>
```

Fixing this incompleteness dropped out naturally from the cleanest code to when covering parametere info for half-written cases such as ``DB.CreateCommand< CMD >``

**Comment by dsyme** [comment]
Subject to CI tests passing, this is now ready.  I've added unit testing for the tooling support.

File: src/fsharp/CompileOps.fs [hunk 1]
```diff
@@ -3478,7 +3477,7 @@ let WriteOptimizationData (tcGlobals, file, ccu,modulInfo) =
 // Abstraction for project reference
 
 type RawFSharpAssemblyDataBackedByFileOnDisk (ilModule: ILModuleDef, ilAssemblyRefs) = 
-    let externalSigAndOptData = ["FSharp.Core";"FSharp.LanguageService.Compiler"]
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: This removes the special treatment of "FSharp.LanguageService.COmpiler" where the sigdata and opdata were being placed into separate files.  There's no need to do this, it doesn't gain us much anyway (it was done to reduce DLL size when we were shipping the Visual F# Tools as a startup component in core Visual Studio, now that it's an optional install we don't need to do this any more, and in any case this only trimmed 1MB of DLL size, which isn't much in context)

File: src/fsharp/CompileOps.fs [hunk 2]
```diff
@@ -4519,17 +4518,19 @@ type TcImports(tcConfigP:TcConfigProvider, initialResolutions:TcAssemblyResoluti
...
+    let asms = 
+        ccuinfos |> List.map  (function
+            | ResolvedImportedAssembly(asm) -> asm
+            | UnresolvedImportedAssembly(assemblyName) -> error(Error(FSComp.SR.buildCouldNotResolveAssemblyRequiredByFile(assemblyName,file),m)))
+
     let g = tcImports.GetTcGlobals()
     let amap = tcImports.GetImportMap()
-    let tcEnv = ccuinfos |> List.fold (fun tcEnv ccuinfo -> Tc.AddCcuToTcEnv(g,amap,m,tcEnv,ccuinfo.FSharpViewOfMetadata,ccuinfo.AssemblyAutoOpenAttributes,false)) tcEnv 
-    tcEnv,(dllinfos,ccuinfos)
+    let tcEnv = (tcEnv, asms) ||> List.fold (fun tcEnv asm -> Tc.AddCcuToTcEnv(g,amap,m,tcEnv,thisAssemblyName,asm.FSharpViewOfMetadata,asm.AssemblyAutoOpenAttributes,asm.AssemblyInternalsVisibleToAttributes))
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: This correctly propagates IVT attributes down for ``#r`` assemblies.  I noticed it when doing unit testing on assemblies using F# Interactive.

File: src/fsharp/CompileOps.fs [hunk 3]
```diff
@@ -4943,29 +4944,16 @@ type LoadClosure with
 //--------------------------------------------------------------------------
 
 /// Build the initial type checking environment
-let GetInitialTcEnv (assemblyName:string option, initm:range, tcConfig:TcConfig, tcImports:TcImports, tcGlobals)  =    
+let GetInitialTcEnv (thisAssemblyName:string, initm:range, tcConfig:TcConfig, tcImports:TcImports, tcGlobals)  =    
     let initm = initm.StartRange
 
-    let internalsAreVisibleHere (asm:ImportedAssembly) =
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: This code is moved to TypeChecker.fs

File: src/fsharp/FSharp.Compiler/InternalsVisibleTo.fs [hunk 4]
```diff
@@ -4,8 +4,6 @@ namespace Microsoft.FSharp
...
-[<assembly:System.Runtime.CompilerServices.InternalsVisibleTo("FSharp.QuickSearch, PublicKey=002400000480000094000000060200000024000052534131000400000100010007D1FA57C4AED9F0A32E84AA0FAEFD0DE9E8FD6AEC8F87FB03766C834C99921EB23BE79AD9D5DCC1DD9AD236132102900B723CF980957FC4E177108FC607774F29E8320E92EA05ECE4E821C0A5EFE8F1645C4C0C93C1AB99285D622CAA652C1DFAD63D745D6F2DE5F17E5EAF0FC4963D261C8A12436518206DC093344D5AD293")>]
-[<assembly:System.Runtime.CompilerServices.InternalsVisibleTo("FSharp.Compiler.Model, PublicKey=002400000480000094000000060200000024000052534131000400000100010007D1FA57C4AED9F0A32E84AA0FAEFD0DE9E8FD6AEC8F87FB03766C834C99921EB23BE79AD9D5DCC1DD9AD236132102900B723CF980957FC4E177108FC607774F29E8320E92EA05ECE4E821C0A5EFE8F1645C4C0C93C1AB99285D622CAA652C1DFAD63D745D6F2DE5F17E5EAF0FC4963D261C8A12436518206DC093344D5AD293")>]
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler/InternalsVisibleTo.fs]: Removing old IVT attributes - these assemblies don't exist.

File: src/fsharp/NameResolution.fs [hunk 5]
```diff
@@ -2905,16 +2914,34 @@ let ResolveCompletionsInType (ncenv: NameResolver) nenv (completionTargets: Reso
...
+                    |> List.map (function Item.Event(FSEvent(_,_,addValRef,removeValRef)) -> [addValRef.LogicalName;removeValRef.LogicalName] | _ -> [])
+                    |> List.concat
+                    |> set
+
+                if addersAndRemovers.IsEmpty then minfos
+                else minfos |> List.filter (fun minfo -> not (addersAndRemovers.Contains minfo.LogicalName))
+
+#if EXTENSIONTYPING
+            // Filter out the ones with mangled names from applying static parameters
+            let minfos =
```
Comment (dsyme) [inline src/fsharp/NameResolution.fs]: This filters the mangled forms of statically-instantiated provided methods from the auto-complete lists

File: src/fsharp/fsc.fs [hunk 6]
```diff
@@ -621,15 +621,13 @@ let EncodeInterfaceData(tcConfig:TcConfig,tcGlobals,exportRemapping,generatedCcu
         let resource = WriteSignatureData (tcConfig,tcGlobals,exportRemapping,generatedCcu,outfile)
         if verbose then dprintf "Generated interface data attribute!\n";
         // REVIEW: need a better test for this
-        let outFileNoExtension = Filename.chopExtension outfile
-        let isCompilerServiceDll = outFileNoExtension.Contains("FSharp.LanguageService.Compiler")
-        if (tcConfig.useOptimizationDataFile || tcGlobals.compilingFslib || isCompilerServiceDll) && not isIncrementalBuild then
```
Comment (dsyme) [inline src/fsharp/fsc.fs]: See earlier comment, this is removing the special treatment of FSharp.LanguageService.Compiler

File: src/fsharp/pars.fsy [hunk 7]
```diff
@@ -3306,6 +3306,8 @@ declExpr:
   | declExpr  INFIX_COMPARE_OP      declExpr { mkSynInfix (rhs parseState 2) $1 $2 $3 }
   | declExpr  DOLLAR                declExpr { mkSynInfix (rhs parseState 2) $1 "$" $3 }
   | declExpr  LESS                  declExpr { mkSynInfix (rhs parseState 2) $1 "<" $3 }
+  | declExpr  LESS                  recover { if not $3 then reportParseErrorAt (rhs parseState 2) (FSComp.SR.parsUnfinishedExpression("<"));  
+                                              exprFromParseError (mkSynInfix (rhs parseState 2) $1 "<" (arbExpr("declExprInfix",(rhs parseState 3).StartRange))) }
```
Comment (dsyme) [inline src/fsharp/pars.fsy]: This new parser recovery rule is needed when processing the incomplete text ``ID<``

File: src/fsharp/vs/ServiceParamInfoLocations.fs [hunk 8]
```diff
@@ -77,66 +77,65 @@ module internal NoteworthyParamInfoLocationsImpl =
...
-            | None ->
-                if AstTraversal.rangeContainsPosEdgesExclusive parenRange pos then
-                    let r = parenRange.Start, [parenRange.End, getNamedParamName synExpr], rpRangeOpt.IsSome
-                    Trace.PrintLine("LanguageServiceParamInfo", fun () -> sprintf "Found single arg paren range %+A from %+A" r expr)
-                    Some r, None
-                else
-                    None, None
-            | _ -> None, None
-
+    let rec searchSynArgExpr traverseSynExpr pos expr =
```
Comment (dsyme) [inline src/fsharp/vs/ServiceParamInfoLocations.fs]: I cleaned up this code a bit to make it more obvious what's going on, allowing me to add the new rule for ``ID<``

File: src/fsharp/vs/service.fs [hunk 9]
```diff
@@ -754,75 +796,68 @@ type TypeCheckInfo
     //
     // The overall aim is to resolve as accurately as possible based on what we know from type inference
     
-    let GetDeclItemsForNamesAtPosition(parseResultsOpt : FSharpParseFileResults option,
```
Comment (dsyme) [inline src/fsharp/vs/service.fs]: I cleaned up some of the code in ``service.fs`` to remove the use of (confusing) inner functions.

File: vsintegration/src/FSharp.VS.FSI/sessions.fs [hunk 10]
```diff
@@ -154,25 +154,37 @@ let catchAll (ie: IEvent<_,_>) : IEvent<_> =
     ie.Add(fun x -> try w(x) with err -> ignore(System.Windows.Forms.MessageBox.Show(err.ToString())); ());
     e
 
-let fsiExeName () = if SessionsProperties.useAnyCpuVersion then "fsianycpu.exe" else "fsi.exe"
-let determineFsiRelativePath () =
+let fsiExeName () = 
+    if SessionsProperties.useAnyCpuVersion then "fsiAnyCPU.exe" else "fsi.exe"
+
+// Use the VS-extension-installed development path if available, relative to the location of this assembly
+let determineFsiRelativePath1 () =
```
Comment (dsyme) [inline vsintegration/src/FSharp.VS.FSI/sessions.fs]: This addition uses the FSI.EXE installed by ``EnableOpenSource.vsix`` if it is available.

File: vsintegration/tests/unittests/Resources.MockTypeProviders/DummyProviderForLanguageServiceTesting/TypeProviderEmit.fsi [hunk 11]
```diff
@@ -1,289 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (dsyme) [inline vsintegration/tests/unittests/Resources.MockTypeProviders/DummyProviderForLanguageServiceTesting/TypeProviderEmit.fsi]: I've replaced the TypeProviderEmit.fs by the latest ``ProvidedTypes.fs`` which has support for static method parameters

**Comment by dsyme** [comment]
The failing tests have been fixed and this is now ready for review an to go in.

**Comment by dsyme** [comment]
I've updated this PR and it is ready to go in.  It would be great if someone could review - maybe @v2m?

**Pull request merged**